Update CI

rahulporuri opened the pull request:
- use etstool module to setup/run tests
- update travis and appveyor config to use the new etstool module
- add relevant scripts to install EDM and run tests

Note that the etstool module added here has been adapted from the
module in pyface, with relevant changes made.

mdickinson commented:
@rahulporuri Thanks for doing this. I haven't had a chance to look at this properly, but I noticed you removing the `numpy` dependency.

While `numpy` should definitely not be a dependency of the packaged traits egg, for CI testing, we should keep the optional dependencies around, else some of the tests get skipped. I think that means we need `numpy`, and ideally also `traitsui` and its dependencies.

rahulporuri commented:
Adding numpy and traitsui as dependencies. One of the reasons why I removed `numpy` as a dependency is because CI now uses enthought eggs for everything instead of using pip, so we end up downloading the massive mkl eggs. There's a noticeable difference in how long the jobs take to finish.

Note that we still have the ability to install packages from pip with the changes introduced in this PR.

mdickinson commented on .travis.yml:
```diff
@@ -1,32 +1,48 @@
-  - pip install --upgrade pip
+  - mkdir -p "${HOME}/.cache/download"
+  - export DISPLAY=:99.0
+  - if [[ ${TRAVIS_OS_NAME} == "linux" ]] ; then sh -e /etc/init.d/xvfb start; fi
```
It's possible that we don't need the framebuffer (though it probably doesn't do any harm).

mdickinson commented on appveyor-clean-cache.txt:
```diff
@@ -0,0 +1 @@
+1
```
Do we need to check this file into the repository? Should we remove it here and add it to `.gitignore` instead?

mdickinson commented on appveyor.yml:
```diff
@@ -1,40 +1,40 @@
 build: false
-shallow_clone: true
+shallow_clone: false
```
Nit: probably easiest just to drop this line.

mdickinson commented on appveyor.yml:
```diff
@@ -1,40 +1,40 @@
 build: false
-shallow_clone: true
+shallow_clone: false
+skip_branch_with_pr: true
```
Another nit: I'm not sure this is all that useful. What we tend to do is disable branch builds altogether through the webhook. (Otherwise, even with this option, I think you still get builds on newly-created branches that don't have an open PR, which is a bit wasteful.)

mdickinson commented on appveyor.yml:
```diff
@@ -1,40 +1,40 @@
-    # See: http://stackoverflow.com/a/13751649/163740
-    CMD_IN_ENV: "cmd /E:ON /V:ON /C .\\appveyor\\run_with_env.cmd"
+    PYTHONUNBUFFERED: "1"
+    INSTALL_EDM_VERSION: "1.9.1"
```
Is it worth adding a comment to the .yml file on why we're not using 1.10?

mdickinson replied:
BTW, Why not 1.9.2?

rahulporuri replied:
added comment and changed edm version to `1.9.2`

mdickinson commented on etstool.py:
```diff
@@ -0,0 +1,288 @@
+    "mock",
+    "nose",
+    "numpy",
+    "traitsui"
```
For CI purposes, I think we need a Qt backend too (for `traits.tests.test_ui_notifiers.TestDecoratorUINotifiers`). PyQt?

mdickinson commented on etstool.py:
```diff
@@ -0,0 +1,288 @@
+dependencies = {
+    "coverage",
+    "cython",
+    "Sphinx",
```
Okay, we do, for the trait documenter tests.

mdickinson commented:
There's something a bit odd going on here. I see this in the test output:
```
test_trait_set_bad (traits.tests.test_get_traits.GetTraitTestCase) ... /Users/mdickinson/.edm/envs/traits-test-3.6/lib/python3.6/unittest/case.py:546: RuntimeWarning: TestResult has no addExpectedFailure method, reporting as passes
  RuntimeWarning)
ok
```

The test in question (`test_trait_set_bad`) _used_ to have an `expectedFailure` decorator, but doesn't any longer, so I'm a bit puzzled.

mdickinson commented:
Using `pip` to do the local install fixes this: it uninstalls the existing `traits` installation.
```
[EXECUTING] edm run -e traits-test-3.6 -- python -m pip install --no-deps .
Processing /Users/mdickinson/Enthought/ETS/traits
Installing collected packages: traits
  Found existing installation: traits 4.6.0
    Uninstalling traits-4.6.0:
      Successfully uninstalled traits-4.6.0
  Running setup.py install for traits ... done
Successfully installed traits-4.7.0.dev124
```

I'll push a fix to this branch shortly.

rahulporuri commented:
I can spend an hour later in the day (my time) addressing your comments and if you're comfortable, you're free to merge tomorrow morning (your time). does that work?

mdickinson commented:
@rahulporuri If you've got the time, that would be great. Yes, please!

rahulporuri commented:
@mdickinson i've addressed most of your comments. Note again that this PR drops support for Python 3.4, which is not available using edm, and we won't be able to support 3.7 (and other newer versions) until edm supports them.

mdickinson commented:
Yes, I noticed the missing 3.4; that's fine with me. We probably want to turn that into an official dropping of 3.4 support sometime soon.

Thanks for the updates!
